subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/annotations/InputFilesPropertyAnnotationHandler.java

```diff
@@ -25,6 +32,18 @@
         return InputFiles.class;
     }
 
+    @Override
+    public ImmutableSet<Class<? extends Annotation>> getAllowedModifiers() {
```

Comment:
Can't we move that method to a superclass or at least extract a constant? One for artifact transform specific input files annotations and one for the others?

Reply:
We have the following sets of categories:

- `@InputArtifact`: `INCREMENTAL`, `NORMALIZATION`
- `@InputArtifactDependencies`: `NORMALIZATION`
- `@InputFile`, `@InputDirectory`, `@InputFiles`: `INCREMENTAL`, `NORMALIZATION`, `OPTIONAL`

We could extract the third set as its reused multiple times, but for that we'd need to define some abstraction that captures the commonalities of those annotations. I don't have a good idea for how to distinguish them against the artifact-related ones. I'd just leave this as it is until we bump into it again.